### ISSUE TITLE
Any change to any bff file causes a full rebuild

### ISSUE THREAD

**Issue by lucasmeijer** [fastbuild/fastbuild#108]
I guess this is by design, but I was wondering what the reasoning is behind the design decision to store the "timestamp of last built"  and the dependencies found  for each node into the serialized depgraph, instead of in a different file.  If kept in a different file, I think you would no longer be in the situation that any .bff file change always causes a rebuild of everything

**Comment by ffulin** [comment]
Hello

The reason a rebuild occurs when editing a bff is not because the dependency/time information is stored in the .db file, but rather because not enough information is currently stored to safely know which build rules are "dirtied" by changes to the bff.  Because of that, we assume all the build rules are dirtied, resulting in a full rebuild.

This is an unfortunate limitation of the current implementation, and is a result of some short sighted implementation choices I made very early on during development.  The good news is I know how to fix this.  The bad news is that it needs some non-trivial internal refactoring.  The more good news though is that I've already started this work - I'm perhaps 50% complete.  The more bad news is that there is still quite a bit to do, and it will be many patches before it's complete.

To summarize, the intended future behaviour is that editing the bff will reparse, but only mark the appropriate parts of the db as dirty, therefore rebuilding only what is needed.

**Comment by uucidl** [comment]
This issue makes using fastbuild on a monorepo quite difficult as any small configuration change generates full rebuilds

**Comment by ffulin** [comment]
@uucidl - I'm not familiar with monorepo - is there something particular about it that makes this more problematic than normal?

**Comment by uucidl** [comment]
It simply increases the frequency and impact because typically in that setup you have more people changing and being affected by changes made to the repository.

Here I'm talking about monorepo with a mother `.bff` at the top. In this case any small reconfiguration (such as someone moving code from a module to another or even adding example code + example targets) would retrigger rebuilds for all developers who resynchronize the repo

**Comment by uucidl** [comment]
Of course `-cache` helps a lot in that scenario!

**Comment by ffulin** [comment]
For the most part you should only setup your bff files so that they only need to be updated when the structure of your projects change (you add new libraries for example).  With the right setup, adding, removing or moving code around should not require editing the bff files.

Having said that, I recognize that some things require editing the bff, and it's annoying to have it rebuild things unnecessarily in those cases.  If the project structure is changing a lot, you might be experiencing the annoyance even more. This limitation will be eliminated in the future - I'm working on it :)

The tip about using the -cache is a good one - it helps a lot for sure.

**Comment by nre-ableton** [comment]
Just wanted to add a +1 to this issue. In our company, FASTBuild is not our canonical build system, but rather, we are using it by way of a generator. This means that our generated `bff` files are subject to semi-frequent changes after `git pull`. Of course, using `-cache` reduces the pain a great deal, but it also would be nice if a full rebuild was not necessary, since even cached builds of our products take a few minutes on my rather beefy workstation.

More significantly, we are considering deploying FASTBuild on our CI servers as well. This is another environment that would almost certainly force a full rebuild every time, which negates a great deal of the CPU savings we'd otherwise get from using FASTBuild there.

**Comment by ffulin** [comment]
> This means that our generated bff files are subject to semi-frequent changes after git pull

@nre-ableton - Is your generator deterministic? (Re-generating the bff won't cause a rebuild if it's identical to before).  Are you listing files explicitly or by wild-card patterns? (Listing them explicitly will cause a rebuild every time you add/remvoe a file, but listing them by pattern will not).

As noted, I hope to have this fixed eventually, but if the two things above are true, it can greatly limit the rebuild pain to trigger only on structural changes to your project which (for most projects) are infrequent.

**Comment by comicfans** [comment]
Also want to add a +1 to this issue.  Recently I'm working on a CMake fastbuild generator, this brings fastbuild power to lots of project ,  but rebuild every time maks fastbuild less useful (ninja build tool didn't have such problem), most CMake project prefer listing file explicitly instead of using pattern.

**Comment by comicfans** [comment]
I've implemented some sort of "minimal rebuild ", even bff re-parse happend,it worked in some simple tests . this add a "SemanticHash" and "SemanticEquals" to Node ,  for executable (any compile/link/library) node,  SemanticHash is calculated by using final running commands, for objectlist , SemanticHash is calculated by all its dependencies (sorted) , when bff change detected, migrate build status from old NodeGraph to new one (first looking up same SemanticHash then test by SemanticEquals).   problem faced during this implementation:
1. FileNode m_Stamp set in DoBuild seems later,  set when construct can avoid some unneeded node built (although node may not run anything because its dependent FileNode's m_Stamp is older ) 
2. old NodeGraph have dynamic dependencies serialized , but new one did't create until actual build, so can not prevserve build status for un-constructed node (ObjectNodes constructed by ObjectListNode). I did manually trigger dynamic dependencies build and it  worked , I'm not sure this is safe way. 
3. some command is constructed at DoBuild time, makes it hard to collect as SemanticHash . and compile command seems can  change at different pass, also may lead different hash  even with same semantic. maybe I can also calculated hash by using node's config variables, but variable seems too many.

**Comment by ffulin** [comment]
@comicfans - As you noted, we really need a mechanism to "hash" the state of a node in order to correctly know when to rebuild. This will be handled by migrating the way nodes are populated to a general reflection system which can generically hash a node so they can be compared between DBs. This work is well underway (v0.93 contains the conversion of Library and ObjectList: two of the biggest ones to port). Additionally, as proof of concept, I have a number of shelved/unsubmitted changes to perform and test this migration for nodes that are converted so far.

Having said all that, this is a huge change to the heart of FASTBuild, and as such I'm being very meticulous about the conversion process. I likely won't port any more nodes for v0.93. It will be several more patches to convert the remaining nodes and robustify the DB migration logic when the DB changes.

The migration to this generic reflection/serialization also takes care of the DynamicDependencies issue you mentioned.

**Comment by comicfans** [comment]
Greate news, hopes to see new version earilear.

**Comment by v1ne** [comment]
@ffulin, is there any way to assist you with these changes to FASTbuild?

At the company I'm working for, I worked around this by globbing the list of input files for the compiler + VCXProj, if possible, in our BFF generator back-end after the generator came up with a concrete list of files. That improved things quite a bit. :-)

**Comment by ffulin** [comment]
@v1ne Hello. Sorry for the incredibly slow reply. 

The ongoing work to re-jig the internals of FASTBuild in order to make bff re-parsing not require a clean build is complicated and hard to get right. I expect that I understand the code better than anyone :), but that didn't stop me already introducing several bugs along the way so far :(.  With that in mind, I don't think it's something that can be helped with for now.

**Comment by n00bmind** [comment]
Hi.
Just to add my 0.02 here..
In our case we have an automated generation step that re-creates our bff everytime. We're trying to be more intelligent about this so that this is only done when strictly necessary..

Anyway what I want to point out is that what this process does is mostly just add/remove files that were added/removed from our source tree. Under no circumstances does this every change our source or output paths, so mostly all object files are there already present in the output folder whenever we compile..

My question is, even in the face of a full rebuild being triggered, whenever each and every source file is being considered again for compilation, couldn't FB just early out the compilation based on a simple timestamp comparison between source-output?
I understand you need a way to unequivocally identify to which source file the obj file emanates from. This could maybe be done by placing some metadata next to every file in the output?

Just a suggestion :)

**Comment by ffulin** [comment]
> couldn't FB just early out the compilation based on a simple timestamp comparison between source-output

@n00bmind - The src/output date stamp comparison is just one thing that needs to be checked when deciding if something needs to rebuild. We also need to consider the command line options, the compiler version and potentially even things like the environment. All of these things are tracked in the database (.fdb), but due to some internal implementation limitations, FASTBuild can't currently preserve this information after re-parsing the a bff.

These internal limitations are being slowly being fixed as part of each new release (my previous comments about the move to a node reflection system), so eventually this problem will be fixed. Eventually, bff changes should not cause any unnecessary rebuilds.

In the interim, the best suggestion I have is, if you are able, to switch your generator to use the .CompilerInputPath and related options to have FASTBuild discover the files, rather than listing them explicitly via .CompilerInputFiles. This will eliminate all reparsing caused by adding/removing files. If you can't do that, then using the cache can be a way to make the rebuilds less painful until this issue is ultimately fixed.

**Comment by n00bmind** [comment]
I see, thanks for the explanation.

That alternative you mention sounds like it could work for us. I'll
investigate further.

**Comment by john-blackburn** [comment]
I'd like to add another suggestion as I have the same problem. I'm currently tinkering with the link part of my BFF (often the most problematic part!) so there is no need for any recompile. But FASTBuild insists on a full recompile of all 4000+ files before it tries my new link-time code! This makes development/debug of the BFF very difficult. I appreciate figuring out which parts of the BFF are "dirty" is very complex but might I suggest the following short term fix? Could we please have a command line option to **completely disable** rebuilding due to changes in the BFF? Of course this is not a long term solution since it will fail to compile many files which ought to be compiled but it would be useful when developing BFF files. The option should be off by default.

Many thanks for your help, FASTBuild is a great tool!

**Comment by ffulin** [comment]
The dev branch now contains the first working version of .db migration: FASTBuild will only rebuild things that are affected by changes to BFF files. (https://github.com/fastbuild/fastbuild/commit/f8d000a5795f5f042b3544a8454bd962d7c74744)

The next version (v0.97) will include this feature, but it will need to be enabled via an explicit toggle ( set AllowDBMigration_Experimental on Settings() ). Once stabilized, this toggle will be removed and this behaviour will be used all the time (most likely in v0.98)

If you are building FASTBuild yourself from source, you can do so from the dev branch to have access to this feature immediately.

**Comment by incursionrob** [comment]
Did that feature make it to v0.98? While testing with v0.98 changing a BFF doesn't seem to rebuild everything, it does however seem to build more that what I would expect. For example a change on how files are blobbed in a project (BFF), makes all projects that it depends on to rebuild, even though there is no change to their BFFs or sources.

**Comment by ffulin** [comment]
@incursionrob .AllowDBMigration_Experimental was added in v0.97. It is off by default and must be enabled by setting AllowDBMigration_Experimental to true in your Settings().

There is at least one very specific known and unresolved bug in this feature (not related to your case) so it's possible there are others. If you have it enabled, please raise a bug for anything that isn't working as expected.

**Comment by incursionrob** [comment]
@ffulin thank you for the quick response, We will try it and let you know if we have any issues.

**Comment by ffulin** [comment]
The DB migration feature has now been promoted from experimental to stable. v0.99 will have DB migration on by default, so changesto bff files will only result in rebuilds for things that have changed. (https://github.com/fastbuild/fastbuild/commit/d9fdc242294c9f98b8d60696bb0dcd1146038f2d). I will leave this issue open until v0.99 is released.

**Comment by uucidl** [comment]
Great news :)

**Comment by ffulin** [comment]
FASTBuild v0.99 has now been released. In this version, the database will persist between builds even if a bff file is edited. Only parts of the build which the bff relates to will rebuild (if any).
